### PR TITLE
Ibm cluster creation

### PR DESCRIPTION
**Pull request by BobbyRadford**
This PR is a followup to the IBM Cloud Scaffolding PR: https://github.com/openshift/installer/pull/4923

It includes additional implementation to support executing `openshift-install create manifests|ignition-configs|cluster`. A full terraform implementation is included for cluster creation and bootstrap destroy. All manifests code has at least been stubbed out to support the IBMCloud platform. The IBM cluster API provider has been pulled in to support Machine/MachineSet provisioning.

Future work will flesh out the remaining implementation.

Enhancement doc: https://github.com/openshift/enhancements/pull/773

Many of the commits are iterative improvements over earlier ones in the PR. If it would be more helpful to squash things down to one commit (excluding mod/vendors) I can certainly do that. Apologies for all the vendor changes! 😬

### CODE REVIEW

**Comment by BobbyRadford** [comment]
Please stand by. I'm working through some rebase issues.

File: pkg/asset/machines/ibmcloud/machines.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import (
+	"fmt"
+
+	ibmcloudprovider "github.com/openshift/cluster-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1beta1"
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/ibmcloud"
+	machineapi "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (jstuever) [inline pkg/asset/machines/ibmcloud/machines.go:12]: nitpick: we usually separate the openshift imports from the others... ie:
```
"fmt"

"github.com/pkg/errors"

"github.com/openshift/installer/pkg/types"
```

**Review by jstuever — CHANGES_REQUESTED** [review]
Overall looks solid. A couple minor tweaks.

**Comment by BobbyRadford** [comment]
@jstuever Thanks for the review. We'll be working on fixes to address your comments this afternoon.

File: data/data/ibmcloud/bootstrap/templates/bootstrap.ign [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        "httpHeaders": [
+          {
+            "name": "Authorization",
+            "value": "${IAM_TOKEN}"
+          }
+        ]
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (BobbyRadford) [reply data/data/ibmcloud/bootstrap/templates/bootstrap.ign:16]: So we are blocked on this by the lack of support for `httpHeaders` in the community TF provider. Opened an issue for adding support. https://github.com/community-terraform-providers/terraform-provider-ignition/issues/16

Can we come back to this in a future PR once the provider has been updated to support the field?

[the remark this answers is not in the document]
Comment (jstuever) [reply data/data/ibmcloud/bootstrap/templates/bootstrap.ign:16]: We can, yes. You mind adding a TODO here?

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 3]
```diff
@@ -151,16 +167,18 @@ func (c *Client) GetDNSZones(ctx context.Context) ([]DNSZoneResponse, error) {
...
 
-		for _, zone := range listZonesResponse.Result {
-			if *zone.Status == "active" {
-				zoneStruct := DNSZoneResponse{
-					Name:            *zone.Name,
-					ID:              *zone.ID,
-					CISInstanceCRN:  *instance.CRN,
-					CISInstanceName: *instance.Name,
-					ResourceGroupID: *instance.ResourceGroupID,
+		if listZonesResponse != nil {
```
[the remark this answers is not in the document]
Comment (BobbyRadford) [reply pkg/asset/installconfig/ibmcloud/client.go]: Done

File: pkg/asset/machines/ibmcloud/machines.go (hunk 1 above)
Comment (BobbyRadford) [reply pkg/asset/machines/ibmcloud/machines.go:10]: Done

File: data/data/ibmcloud/bootstrap/templates/bootstrap.ign (hunk 2 above)
[the remark this answers is not in the document]
Comment (BobbyRadford) [reply data/data/ibmcloud/bootstrap/templates/bootstrap.ign:16]: Sure. I can do that.

**Comment by BobbyRadford** [comment]
@jstuever Anything left for us to touch up?

**Pull request merged**